V2 all post

VimS opened the pull request:
Nach interner Besprechung wollten wir den Request an DeepL halbwegs gerade ziehen.
Deswegen wird nun bis auf den Api-Key alles im Post-Body mitgeschickt.

VimS committed:
- remove buildUrl
 - remove buildBody
 - write params to post-body with http_build_query instead
 - remove Tests

VimS committed:
- remove unnecessary constants

VimS committed:
- extract buildQuery from translate

VimS committed:
- code cleanup
 - fix Bugs discovered while writing Tests
 - add Testcases for testBuildQuery

VimS committed:
- add Testcase for multiple Texts as Array
- code cleanup
- readd functionality for text as array
- try to reduce CyclomaticComplexity and NPathComplexity

VimS committed:
- further cleanup
- removeEmptyParams to reduce complexity of buildQuery
- add Testcases for removeEmptyParams

VimS committed:
- remove PHP_QUERY_RFC3986 from due to PHP 5.3

denex23 commented on src/DeepL.php:
```diff
@@ -181,6 +126,92 @@ public function __destruct()
+     *
+     * @throws DeepLException
+     */
+    protected function request($url, $body)
```
parameter $body is missing in docblock

VimS replied:
Was added to docblock

VimS committed:
- add a little documentation for a special Case

denex23 commented on src/DeepL.php:
```diff
@@ -281,194 +342,34 @@ protected function checkLanguages($sourceLanguage, $destinationLanguage)
-        $outlineDetection = 1,
-        array $splittingTags = null
+    protected function buildQuery(
+        $paramsArray
```
code style: line breaks aren't necessary anymore

VimS replied:
cleaned it up

denex23 commented on src/DeepL.php:
```diff
@@ -281,194 +342,34 @@ protected function checkLanguages($sourceLanguage, $destinationLanguage)
-            if ($first) {
-                $first = false;
+        if (true === is_array($paramsArray['text'])) {
+            $text = $paramsArray['text'];
+            unset($paramsArray['text']);
+            $textString = '';
+            foreach ($text as $textElement) {
+                $textString .= '&text='.rawurlencode($textElement);
```
```suggestion
            $textString = '';
            foreach ($paramsArray['text'] as $textElement) {
                $textString .= '&text='.rawurlencode($textElement);
            }
            $paramsArray['text'] = $textString;
```
the if clause at the end would be unnecessary also.

VimS replied:
That would lead for the textstring to get double encoded and mixed up:
&text=text&text=more%20text&text=even%20more%20text
would become:
text=%26text%3Dtext%26text%3Dmore%2520text%26text%3Deven%2520more%2520text

denex23 replied:
Your're right. I didn't consider the http_build_query function. My fault.

denex23 reviewed (commented):
Just a few improvements, nothing that would lead to an error.

VimS committed:
- fix Docblock and Codestyle

VimS committed:
- add Testcase for Handling of html5-Tags

VimS committed:
- fix commemt for Test

VimS committed:
- remove unessesarry $resource-parameter from translate-function

VimS committed:
- fix Tests for translate()

VimS committed:
Breaking-Changes:
- Remove string-return of translate()
- Add Exception when $tagHandling is send as array instead of String
- Adjust and Add Testcases

VimS committed:
- remove checkLanguages with Tests

VimS committed:
- add Testcase for unsupported Languages

VimS committed:
- use ReflectionMethod insted of array

VimS committed:
- revert from ReflectionMethod to Array, because MessDetection did not like it.

VimS committed:
- remove unused Class

denex23 commented on tests/integration/DeepLApiTest.php:
```diff
@@ -317,4 +317,34 @@ public function testTranslateWithHTML()
+    }
+
+    /**
+     * Test translate() with unsupported sourceLanguage
```
```suggestion
     * Test translate() with unsupported targetLanguage
```

VimS committed:
Update tests/integration/DeepLApiTest.php

Co-authored-by: Dennis Exner <59438491+dexner@users.noreply.github.com>

VimS committed:
some minor improvements:
- sort class properties by type and alphabetically
- change parametes of request-Function to get rid of unnecessary variable initialization

VimS committed:
- Update README.md
- add source & target option to languages-function
- add Testscases

VimS committed:
- Updated README.md

VimS committed:
- Updated README.md

VimS committed:
- Updated README.md: User Flags of countries which are actually supported by DeepL

VimS committed:
- add Changelog.md: tried my best to add Changes done in previous Versions

VimS committed:
- add link to CHANGELOG.md in README.md
